clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -0,0 +1,142 @@
+:email:             paul.m.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently, WildFly supports 2 mechanisms for persistent EJB timers: a file-based implementation and a JDBC-based implementation.
+Of these, only the JDBC-based implementation is suitable for use in a cluster.   However, this implementation does not scale well to large to medium sized clusters and relies on a single point of failure.
+WildFly needs an persistent timer service implementation suitable for use in a cluster is both highly-available and reasonably scalable.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-7628[WFLY-7628]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-7628[WFLY-7628]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:msvehla@redhat.com[Martin Svehla]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [x] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* Introduce a proper SPI for a TimerService within org.jboss.as.ejb3.timerservice.spi, to facilitate multiple implementations.
+** Adapt the existing TimerResource runtime attributes/operations to use the TimerService SPI instead of referencing specific implementation classes.
+** Adapt existing TimerService implementation to this SPI.
+* Add a CompositeTimerService implementation that delegates to separate persistent and transient TimerService implementations.
+* Implement a distributed implementation of the TimerService SPI.
+** Add a org.wildfly.clustering.ejb.timer SPI to the wildfly-clustering-ejb-spi module to facilitate multiple distributed timer manager implementations.
+** Add a distributed TimerService implementation to the EJB subsystem that delegates to the wildfly-clustering-ejb-spi for timer management and scheduling.
+** Add an Infinispan implementation of the org.wildfly.clustering.ejb.timer API that assigns execution ownership of timers to cluster members based on consistent hashing of the timer identifier.
+*** Leverage the existing Scheduler implementation within wildfly-clustering-ee-cache for timer scheduling
+*** Leverage the existing PrimaryOwnerScheduler decorator that uses consistent hashing of the timer ID to distribute timer execution among cluster members
+* Add a management resource and capability to the distributable-ejb subsystem to configure the Infinispan TimerManagement implementation.
```

Comment:
I had a vague idea to change timer id from string to int (WFLY-15568). Is this doable in infinispan-based distributed timer service impl? The switch will be a long-term consideration, but it will be good to have the flexibility for the potential change of id type. I expect int id will have meaningful performance advantage in both database and infinispan-based timer service impl.

Reply:
The infinispan-based timer service currently uses UUIDs, as these need to be unique across cluster members (rather than a counter).  Those UUIDs are then converted to strings for use by the management API.